aten/src/ATen/native/NegateFallback.cpp

```diff
@@ -0,0 +1,122 @@
+  //  Materialize other inputs as in (1).
+  //
+  //  It is important to be able to tell if we READ from an argument and if we
+  //  WRITE from an argument.  Conservative approach is to assume that we always
+  //  READ from an argument, but in out-of-place operations you can skip
+  //  negating inputs on entry that never get used.  In current schema we
+  //  can't easily tell if inplace situation has happened, so don't do it.
+
+  //  std::cerr << "neg fallback " << op.schema().name() << "\n";
+
+  const auto& arguments = op.schema().arguments();
+  const auto num_arguments = arguments.size();
+  const auto stack_start = stack->size() - num_arguments;
+
+  c10::optional<bool> is_write;
+  for (int64_t i = 0; i < num_arguments; ++i) {
+    const auto& alias_info = arguments[i].alias_info();
+    if (alias_info.has_value()) {
+      if (is_write.has_value()) {
+        TORCH_CHECK(*is_write == alias_info->isWrite(),
+          "Unsupported operator for negation fallback: ", op.schema().name(),
+          "Negation fallback doesn't work for operators with a mix "
+          "mutable and non-mutable inputs that alias with outputs, "
+           "this must be implemented manually.  "
+          "If you got this error on a core op, please report a bug to PyTorch.");
+      } else {
+        is_write = alias_info->isWrite();
+      }
+    }
+  }
+
+  if (is_write.has_value() && !*is_write) {
+    // We assume that view operators automatically handle negation
+    // correctly by propagating the Negative dispatch key in key_set.
+    // This is not necessarily always right, so you should test these cases.
+    op.redispatchBoxed(dispatch_keys & c10::DispatchKeySet(DispatchKeySet::FULL_AFTER, DispatchKey::Negative), stack);
+    return;
+  }
+
+  // Mutable inputs to be tracked separately
+  std::vector<Tensor> mutable_inputs;
+
+  for (int64_t i = 0; i < num_arguments; ++i) {
+    auto& ivalue = (*stack)[stack_start + i];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& argument = arguments[i];
+    bool mut_arg = false;
+    if (argument.alias_info()) {
+      // Was already tested by is_write loop above
+      TORCH_INTERNAL_ASSERT_DEBUG_ONLY(argument.alias_info()->isWrite());
+      mut_arg = true;
+    }
+    auto* impl = ivalue.unsafeToTensorImpl();
+    if (!impl->is_neg()) {
+      continue;
+    }
+
+    auto tensor = std::move(ivalue).toTensor();
+    if (mut_arg) {
+      // TODO: This is a waste if the argument is write only
+      native::neg_(tensor);
+      tensor.set_neg(false);
+      mutable_inputs.emplace_back(tensor);
+    } else {
+      tensor = native::resolve_neg(tensor);
+    }
+    (*stack)[stack_start + i] = std::move(tensor);
+  }
+
+  op.redispatchBoxed(dispatch_keys & c10::DispatchKeySet(DispatchKeySet::FULL_AFTER, DispatchKey::Negative), stack);
+
+  for (auto& mutable_input : mutable_inputs) {
+    native::neg_(mutable_input);
+    mutable_input.set_neg(true);
+  }
+}
```

Comment:
As the logic here is substantively the same, it would be nice if we could reuse the implementation in the two cases. Maybe you can template over the functions you need to call (e.g., `resolve_neg` and `set_neg` and `neg_`) and then instantiate it twice. Maybe a class you can inherit from and then fill in the meaning of some functions would work OK too.